// neo4j-driver/lib/v1/internal/logger.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _levels; /**
              * Copyright (c) 2002-2018 "Neo4j,"
              * Neo4j Sweden AB [http://neo4j.com]
              *
              * This file is part of Neo4j.
              *
              * Licensed under the Apache License, Version 2.0 (the "License");
              * you may not use this file except in compliance with the License.
              * You may obtain a copy of the License at
              *
              *     http://www.apache.org/licenses/LICENSE-2.0
              *
              * Unless required by applicable law or agreed to in writing, software
              * distributed under the License is distributed on an "AS IS" BASIS,
              * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
              * See the License for the specific language governing permissions and
              * limitations under the License.
              */


var _error = require('../error');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ERROR = 'error';
var WARN = 'warn';
var INFO = 'info';
var DEBUG = 'debug';

var DEFAULT_LEVEL = INFO;

var levels = (_levels = {}, (0, _defineProperty3.default)(_levels, ERROR, 0), (0, _defineProperty3.default)(_levels, WARN, 1), (0, _defineProperty3.default)(_levels, INFO, 2), (0, _defineProperty3.default)(_levels, DEBUG, 3), _levels);

/**
 * Logger used by the driver to notify about various internal events. Single logger should be used per driver.
 */

var Logger = function () {

  /**
   * @constructor
   * @param {string} level the enabled logging level.
   * @param {function(level: string, message: string)} loggerFunction the function to write the log level and message.
   */
  function Logger(level, loggerFunction) {
    (0, _classCallCheck3.default)(this, Logger);

    this._level = level;
    this._loggerFunction = loggerFunction;
  }

  /**
   * Create a new logger based on the given driver configuration.
   * @param {object} driverConfig the driver configuration as supplied by the user.
   * @return {Logger} a new logger instance or a no-op logger when not configured.
   */


  (0, _createClass3.default)(Logger, [{
    key: 'isErrorEnabled',


    /**
     * Check if error logging is enabled, i.e. it is not a no-op implementation.
     * @return {boolean} `true` when enabled, `false` otherwise.
     */
    value: function isErrorEnabled() {
      return isLevelEnabled(this._level, ERROR);
    }

    /**
     * Log an error message.
     * @param {string} message the message to log.
     */

  }, {
    key: 'error',
    value: function error(message) {
      if (this.isErrorEnabled()) {
        this._loggerFunction(ERROR, message);
      }
    }

    /**
     * Check if warn logging is enabled, i.e. it is not a no-op implementation.
     * @return {boolean} `true` when enabled, `false` otherwise.
     */

  }, {
    key: 'isWarnEnabled',
    value: function isWarnEnabled() {
      return isLevelEnabled(this._level, WARN);
    }

    /**
     * Log an warning message.
     * @param {string} message the message to log.
     */

  }, {
    key: 'warn',
    value: function warn(message) {
      if (this.isWarnEnabled()) {
        this._loggerFunction(WARN, message);
      }
    }

    /**
     * Check if info logging is enabled, i.e. it is not a no-op implementation.
     * @return {boolean} `true` when enabled, `false` otherwise.
     */

  }, {
    key: 'isInfoEnabled',
    value: function isInfoEnabled() {
      return isLevelEnabled(this._level, INFO);
    }

    /**
     * Log an info message.
     * @param {string} message the message to log.
     */

  }, {
    key: 'info',
    value: function info(message) {
      if (this.isInfoEnabled()) {
        this._loggerFunction(INFO, message);
      }
    }

    /**
     * Check if debug logging is enabled, i.e. it is not a no-op implementation.
     * @return {boolean} `true` when enabled, `false` otherwise.
     */

  }, {
    key: 'isDebugEnabled',
    value: function isDebugEnabled() {
      return isLevelEnabled(this._level, DEBUG);
    }

    /**
     * Log a debug message.
     * @param {string} message the message to log.
     */

  }, {
    key: 'debug',
    value: function debug(message) {
      if (this.isDebugEnabled()) {
        this._loggerFunction(DEBUG, message);
      }
    }
  }], [{
    key: 'create',
    value: function create(driverConfig) {
      if (driverConfig && driverConfig.logging) {
        var loggingConfig = driverConfig.logging;
        var level = extractConfiguredLevel(loggingConfig);
        var loggerFunction = extractConfiguredLogger(loggingConfig);
        return new Logger(level, loggerFunction);
      }
      return this.noOp();
    }

    /**
     * Create a no-op logger implementation.
     * @return {Logger} the no-op logger implementation.
     */

  }, {
    key: 'noOp',
    value: function noOp() {
      return noOpLogger;
    }
  }]);
  return Logger;
}();

var NoOpLogger = function (_Logger) {
  (0, _inherits3.default)(NoOpLogger, _Logger);

  function NoOpLogger() {
    (0, _classCallCheck3.default)(this, NoOpLogger);
    return (0, _possibleConstructorReturn3.default)(this, (NoOpLogger.__proto__ || (0, _getPrototypeOf2.default)(NoOpLogger)).call(this, null, null));
  }

  (0, _createClass3.default)(NoOpLogger, [{
    key: 'isErrorEnabled',
    value: function isErrorEnabled() {
      return false;
    }
  }, {
    key: 'error',
    value: function error(message) {}
  }, {
    key: 'isWarnEnabled',
    value: function isWarnEnabled() {
      return false;
    }
  }, {
    key: 'warn',
    value: function warn(message) {}
  }, {
    key: 'isInfoEnabled',
    value: function isInfoEnabled() {
      return false;
    }
  }, {
    key: 'info',
    value: function info(message) {}
  }, {
    key: 'isDebugEnabled',
    value: function isDebugEnabled() {
      return false;
    }
  }, {
    key: 'debug',
    value: function debug(message) {}
  }]);
  return NoOpLogger;
}(Logger);

var noOpLogger = new NoOpLogger();

/**
 * Check if the given logging level is enabled.
 * @param {string} configuredLevel the configured level.
 * @param {string} targetLevel the level to check.
 * @return {boolean} value of `true` when enabled, `false` otherwise.
 */
function isLevelEnabled(configuredLevel, targetLevel) {
  return levels[configuredLevel] >= levels[targetLevel];
}

/**
 * Extract the configured logging level from the driver's logging configuration.
 * @param {object} loggingConfig the logging configuration.
 * @return {string} the configured log level or default when none configured.
 */
function extractConfiguredLevel(loggingConfig) {
  if (loggingConfig && loggingConfig.level) {
    var configuredLevel = loggingConfig.level;
    var value = levels[configuredLevel];
    if (!value && value !== 0) {
      throw (0, _error.newError)('Illegal logging level: ' + configuredLevel + '. Supported levels are: ' + (0, _keys2.default)(levels));
    }
    return configuredLevel;
  }
  return DEFAULT_LEVEL;
}

/**
 * Extract the configured logger function from the driver's logging configuration.
 * @param {object} loggingConfig the logging configuration.
 * @return {function(level: string, message: string)} the configured logging function.
 */
function extractConfiguredLogger(loggingConfig) {
  if (loggingConfig && loggingConfig.logger) {
    var configuredLogger = loggingConfig.logger;
    if (configuredLogger && typeof configuredLogger === 'function') {
      return configuredLogger;
    }
  }
  throw (0, _error.newError)('Illegal logger function: ' + loggingConfig.logger);
}

exports.default = Logger;
